pkg/featurechangestopper/featurechangestopper.go

```diff
@@ -0,0 +1,140 @@
+			c.queue.Add("cluster")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.queue.Add("cluster")
+		},
+	})
+
+	return c
+}
+
+func (c *TechPreviewChangeStopper) syncHandler(ctx context.Context, shutdownFn context.CancelFunc) error {
+	featureGates, err := c.featureGateLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+	techPreviewNowSet := featureGates.Spec.FeatureSet == configv1.TechPreviewNoUpgrade
+	if techPreviewNowSet != c.startingTechPreviewState {
+		klog.Infof("TechPreviewNoUpgrade status changed from %v to %v, shutting down.", c.startingTechPreviewState, techPreviewNowSet)
+		if shutdownFn != nil {
+			shutdownFn()
+		}
+		return nil
+	}
+	return nil
+}
+
+func (c *TechPreviewChangeStopper) Run(runContext context.Context, shutdownFn context.CancelFunc) {
+	// don't let panics crash the process
+	defer utilruntime.HandleCrash()
+	// make sure the work queue is shutdown which will trigger workers to end
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting stop-on-techpreview-change controller")
+
+	// wait for your secondary caches to fill before starting your work
+	if !cache.WaitForCacheSync(runContext.Done(), c.cacheSynced...) {
+		return
+	}
+
+	// runWorker will loop until "something bad" happens.  The .Until will
+	// then rekick the worker after one second
+	go wait.UntilWithContext(runContext, func(ctx context.Context) {
+		c.runWorker(ctx, shutdownFn)
+	}, time.Second)
+
+	// wait until we're told to stop
+	<-runContext.Done()
+	klog.Infof("Shutting down stop-on-techpreview-change controller")
+}
+
+func (c *TechPreviewChangeStopper) runWorker(ctx context.Context, shutdownFn context.CancelFunc) {
+	// hot loop until we're told to stop.  processNextWorkItem will
+	// automatically wait until there's work available, so we don't worry
+	// about secondary waits
+	for c.processNextWorkItem(ctx, shutdownFn) {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false
+// when it's time to quit.
+func (c *TechPreviewChangeStopper) processNextWorkItem(ctx context.Context, shutdownFn context.CancelFunc) bool {
+	// pull the next work item from queue.  It should be a key we use to lookup
+	// something in a cache
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of
+	// work
+	defer c.queue.Done(key)
+
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := c.syncHandler(ctx, shutdownFn)
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your
+		// key. This will reset things like failure counts for per-item rate
+		// limiting
+		c.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for
+	// pluggable error handling which can be used for things like
+	// cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
```

Comment:
nit: Formatting an `error` with `Errorf`, might as well [use `%w` instead of `%v` for compat with `Unwrap`][1].

[1]: https://pkg.go.dev/fmt#Errorf

Reply:
Floating upstream with kubernetes/community#6266